### PR TITLE
Add initial version of PyTorchTrainer in orca.learn

### PR DESCRIPTION
**Pull request by hkvision**

### CODE REVIEW

**Comment by hkvision** [comment]
@jason-dai Is this what you expect?

File: pyzoo/zoo/orca/learn/torch/__init__.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/torch/__init__.py:1]: `orca.learn.pytorch`

File: pyzoo/zoo/orca/learn/torch/torch_trainer.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from ray.util.sgd.torch import TorchTrainer as RaySGDTorchTrainer
+
+
+class TorchTrainer(object):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/torch/torch_trainer.py:20]: PyTorchTrainer

File: pyzoo/zoo/orca/learn/torch/__init__.py (hunk 1 above)
Comment (hkvision) [reply pyzoo/zoo/orca/learn/torch/__init__.py:1]: In TorchNet, the package is torch not pytorch, shouldn't we keep aligned?

Comment (jason-dai) [reply pyzoo/zoo/orca/learn/torch/__init__.py:1]: let's make it explicit that this is for pythorch

Comment (hkvision) [reply pyzoo/zoo/orca/learn/torch/__init__.py:1]: sure!

File: pyzoo/zoo/orca/learn/pytorch/pytorch_trainer.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+class PyTorchTrainer(object):
+    def __init__(self, model_creator, data_creator, optimizer_creator,
+                 loss_creator=None, scheduler_creator=None, training_operator_cls=None,
+                 initialization_hook=None, config=None, num_workers=1,
+                 use_fp16=False, use_tqdm=False,
+                 scheduler_step_freq="batch", num_replicas=None,
+                 batch_size=None, data_loader_args=None):
+        # Lift TorchTrainer to an Actor so that its local worker would be
+        # created on the cluster as well.
+        RemoteTrainer = ray.remote(TorchTrainer)
```
Comment (hkvision) [inline pyzoo/zoo/orca/learn/pytorch/pytorch_trainer.py:30]: @jason-dai In our yarn settings, lifting the original TorchTrainer as an Actor would be better. See this issue for more discussion: https://github.com/ray-project/ray/issues/8336

**Comment by hkvision** [comment]
@jason-dai I add a ut. Take a further look. Would this pr be good for an initial version?

File: pyzoo/zoo/orca/learn/pytorch/pytorch_trainer.py (hunk 3 above)
Comment (jason-dai) [reply pyzoo/zoo/orca/learn/pytorch/pytorch_trainer.py:28]: what if we are running on local?

**Comment by hkvision** [comment]
jenkins passed: http://10.239.47.210:18888/job/ZOO-PR-Validation/3427/

File: pyzoo/zoo/orca/learn/pytorch/pytorch_trainer.py (hunk 3 above)
Comment (hkvision) [reply pyzoo/zoo/orca/learn/pytorch/pytorch_trainer.py:28]: We can do this, but there's some concerns:
- Normally driver would start with 0 cores, we need to ask users to explicitly specify how many cores for the driver.
- Sometimes the driver may not have enough resources?
- If the driver hardware is worse than workers, the driver will be the bottleneck for the distributed training.

Do you want to keep the local driver?
If so, I would test this. Previously I test the remote way.

Comment (hkvision) [reply pyzoo/zoo/orca/learn/pytorch/pytorch_trainer.py:28]: > what if we are running on local?

Has no problem with ray local. The unit test covers :)

**Comment by jason-dai** [comment]
LGTM

**Pull request merged**